File: Scanner.py
from flask import Flask, jsonify, request
from flask_cors import CORS
import requests
from bs4 import BeautifulSoup
from urllib.parse import urlparse
import ssl
import socket
import psycopg2
import redis
import json
from playwright.sync_api import sync_playwright

app = Flask(__name__)
CORS(app)


cache = redis.Redis(host='localhost', port=6379, db=0)


def get_conn():
    return psycopg2.connect(
        dbname="scanner",
        user="postgres",
        password="kantero@09",
        host="localhost",
        port="5432"
    )


SECURITY_HEADERS = [
    'Content-Security-Policy',
    'Strict-Transport-Security',
    'X-Content-Type-Options',
    'X-Frame-Options',
    'Referrer-Policy'
]


def check_ssl(hostname):
    try:
        context = ssl.create_default_context()
        with socket.create_connection((hostname, 443), timeout=5) as sock:
            with context.wrap_socket(sock, server_hostname=hostname) as ssock:
                cert = ssock.getpeercert()
                return {
                    "valid": True,
                    "issuer": str(cert.get('issuer'))
                }
    except Exception as e:
        return {
            "valid": False,
            "error": str(e)
        }

HEADER_VULNS = {
    'Content-Security-Policy': 'Posible XSS y carga de scripts maliciosos',
    'Strict-Transport-Security': 'Posible downgrade a HTTP inseguro',
    'X-Content-Type-Options': 'Posible MIME sniffing',
    'X-Frame-Options': 'Posible Clickjacking',
    'Referrer-Policy': 'Filtración de información sensible',
    'Permissions-Policy': 'Abuso de permisos del navegador'
}


def check_headers(headers):
    results = []

    for h in SECURITY_HEADERS:
        exists = h in headers

        if exists:
            results.append({
                "header": h,
                "status": "OK",
                "message": "Sin vulnerabilidades detectadas"
            })
        else:
            results.append({
                "header": h,
                "status": "FALTA",
                "message": HEADER_VULNS.get(h, "Riesgo desconocido")
            })

    return results


def analyze_cookies(response):
    results = []

    for cookie in response.cookies:
        results.append({
            "name": cookie.name,
            "secure": cookie.secure,
            "httpOnly": cookie._rest.get('HttpOnly', False)
        })

    return results


def analyze_forms(html):
    soup = BeautifulSoup(html, 'html.parser')

    forms = soup.find_all('form')

    results = []

    for i, form in enumerate(forms, start=1):

        inputs = form.find_all('input')

        csrf_found = any(
            'csrf' in (inp.get('name') or '').lower() or
            'token' in (inp.get('name') or '').lower()
            for inp in inputs
        )

        results.append({
            "form": i,
            "inputs": len(inputs),
            "csrf": csrf_found
        })

    return results


def save_scan(data):
    try:
        conn = get_conn()
        cur = conn.cursor()
        cur.execute(
            "INSERT INTO scans (url, status_code, server) VALUES (%s, %s, %s)",
            (data["url"], data["status_code"], data["server"])
        )
        conn.commit()
        cur.close()
        conn.close()
        print(">>> GUARDADO EN POSTGRESQL <<<")
    except Exception as e:
        print("Error guardando en PostgreSQL:", e)


def scan(url):

    cached = cache.get(url)
    if cached:
        print(">>> RESPUESTA DESDE REDIS <<<")
        return json.loads(cached.decode())

    print(">>> ESCANEANDO URL (BROWSER MODE) <<<")

    try:
        r = requests.get(url, timeout=10)
        status = r.status_code
        server = r.headers.get("Server", "Desconocido")
        headers = check_headers(r.headers)
    except:
        status = None
        server = None
        headers = None

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()
        page.goto(url, timeout=10000)
        html = page.content()
        cookies = page.context.cookies()
        browser.close()

    parsed = urlparse(url)
    ssl_data = check_ssl(parsed.hostname)

    data = {
        "url": url,
        "status_code": status,
        "server": server,
        "headers": headers,
        "cookies": cookies,
        "forms": analyze_forms(html),
        "ssl": ssl_data
    }

    cache.setex(url, 60, json.dumps(data))
    save_scan(data)

    return data



@app.route('/scan')
def web_scan():
    url = request.args.get('url')

    if not url:
        return jsonify({"error": "No URL provided"})

    if not url.startswith("http"):
        url = "https://" + url

    result = scan(url)
    return jsonify(result)


if __name__ == '__main__':
    app.run(debug=True)

File: Scanner.js
async function scanWebsite() {

    const url = document.getElementById("urlInput").value
    const results = document.getElementById("results")

    results.innerHTML = "<p>Escaneando...</p>"

    try {

        const response = await fetch(
            `http://127.0.0.1:5000/scan?url=${encodeURIComponent(url)}`
        )

        const data = await response.json()

        if (data.error) {

            results.innerHTML = `
                <div class="result">
                    <h2>Error</h2>
                    <p>${data.error}</p>
                </div>
            `
            return
        }

        let html = `
            <div class="result">
                <h2>${data.url}</h2>
                <p><strong>Status:</strong> ${data.status_code}</p>
                <p><strong>Servidor:</strong> ${data.server}</p>
            </div>
        `

        html += `
            <div class="result">
                <h2>Headers de Seguridad</h2>
        `

        if (data.headers && Array.isArray(data.headers)) {

            data.headers.forEach(header => {

                let solution = ""

                if (header.header === "Content-Security-Policy") {
                    solution = "Añadir una política CSP para bloquear scripts maliciosos."
                }

                else if (header.header === "Strict-Transport-Security") {
                    solution = "Forzar HTTPS usando HSTS."
                }

                else if (header.header === "X-Content-Type-Options") {
                    solution = "Configurar X-Content-Type-Options en nosniff."
                }

                else if (header.header === "X-Frame-Options") {
                    solution = "Configurar X-Frame-Options en DENY o SAMEORIGIN."
                }

                else if (header.header === "Referrer-Policy") {
                    solution = "Limitar información enviada en el encabezado Referer."
                }

                else if (header.header === "Permissions-Policy") {
                    solution = "Restringir permisos sensibles del navegador."
                }

                html += `
                    <div class="result">

                        <p>
                            <strong>${header.header}</strong>
                        </p>

                        <p>
                            Estado:
                            <strong>${header.status}</strong>
                        </p>

                        <p>
                            Mensaje:
                            ${header.message}
                        </p>

                        <p>
                            Solución:
                            ${solution}
                        </p>

                    </div>
                `
            })

        } else {
            html += `<p>No se pudieron obtener headers</p>`
        }

        html += `</div>`

        html += `
            <div class="result">
                <h2>Cookies</h2>
        `

        if (data.cookies.length === 0) {

            html += `<p>No se detectaron cookies</p>`

        } else {

            data.cookies.forEach(cookie => {

                html += `
                    <p>
                        ${cookie.name}
                        | Secure: ${cookie.secure}
                        | HttpOnly: ${cookie.httpOnly}
                    </p>
                `
            })
        }

        html += `</div>`

        html += `
            <div class="result">
                <h2>Formularios</h2>
        `

        if (data.forms.length === 0) {

            html += `<p>No se encontraron formularios</p>`

        } else {

            data.forms.forEach(form => {

                html += `
                    <p>
                        Formulario #${form.form}
                        | CSRF: ${form.csrf}
                    </p>
                `
            })
        }

        html += `</div>`

        html += `
            <div class="result">
                <h2>SSL</h2>
                <p>Válido: ${data.ssl.valid}</p>
            </div>
        `

        results.innerHTML = html

    } catch (error) {

        results.innerHTML = `
            <div class="result">
                <h2>Error</h2>
                <p>${error}</p>
            </div>
        `
    }
}